### ISSUE TITLE
[aim for beginning of q1] Accessibility: Most/All Pages in Precal (maybe other books) Jump from H1 to H3

### ISSUE THREAD

**Issue by lwtchu** [openstax/cnx-recipes#5304]
ORIGINAL ISSUE: https://github.com/openstax/unified/issues/2402 

# Description

This was submitted by an Instructional Accessibility Specialist at Ohlone College who may want to adopt our books. This likely applies to most or all books in REX.

Unsure if this would be fixed by @kswanson33 & co but since the issue occurs on REX, I opened in the Unified repo. 

"...the Precalculus 2e textbook has one reoccurring accessibility challenge. The pages jump from the main title a heading 1 level to the next heading tagged as a heading level 3. The skipped heading doesn't follow a logical sequence and may confuse Screen Reader users. (see the attached image)"

**To Reproduce**
Steps to reproduce the behavior are
1. Go to https://openstax.org/books/precalculus-2e/pages/1-1-functions-and-function-notation 
2. Navigating between the title "1.1 Functions and Function Notation" and "Learning Objectives", note the jump from H1 to H3

**Expected behavior**
"Learning Objectives" is expected to be an H2

**Screenshots**
<img width="271" alt="Screenshot PreCalculus 2e" src="https://github.com/openstax/unified/assets/7863733/af5e3b96-ce7f-4bee-8a3d-217fad6d5898">

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

# Acceptance Tests
Additional acceptance tests that are required by this bug fix.

1. Check below a/c in polish/spanish books if applicable
2.
3. 

**title**: template  
**categories**: Interactive Component, Navigation, TOC, Math, Book Content, Other
> GIVEN something  
> AND something else  
> WHEN something  
> AND something else  
> THEN something  
> AND something else  

# Pipeline
1A-PRE DESIGN
- [ ] PDM approves UX LoFi Design

1C-DESIGN
  - [ ] PDM approves UX HiFi Design

2B-REVIEW
  - [ ] UX approves change
  - [ ] QA approves change
  - [ ] PDM approves change
  
3A-PEND RELEASE
  - [ ] QA verifies that the acceptance tests are imported into TestRail

# Process 
[Work Management for Unified](https://github.com/openstax/unified/blob/master/CONTRIBUTING.md)

@openstaxalina commented:
This is happening because REX removes the main header which is an h2 in Archive. This might be because the h1 is the book title, page titles are expected to be h2s, and therefore the learning objectives header would be an h3.

Calculus, U.S. History, and Writing Guide are all behaving the same way, so this is likely for all books.

@openstaxalina commented: 
per @philschatz we may be able to make Learning Objectives an h2...but this is a fair amount of testing and work, so we want to prioritize this against other accessibility issues. Also wait for Level Access to assist with prioritization. @erikayvonne adding to the a11y issues spreadsheet.

@TomWoodward commented:
i'm also seeing the heading structure starting in different places, like this composite page starting at h4 https://openstax.org/books/astronomy/pages/1-for-further-exploration

probably because the composite content was defined within one of the section modules and then moved right? so the heading structure made sense when it was written but not after it was moved

in any case i don't think it makes sense for rex to have to re-map all the headings to correct their order, it seems like the content produced from baking should be correct to begin with, but i'm not sure how to align it so that it plays nicely with the pdf.

@openstaxalina commented:
Thank you, @TomWoodward. Like you suggested, we are concerned that if we change the nesting level of the content, this will "break"/cause the wrong header levels in the PDF.

@erikayvonne would it be possible to get priority/severity info on this issue from Level Access (if we don't have it already)?

One idea: once we have the separate web baking step, we could investigate whether to bump all the headers up 1-2 levels specifically for web.

@TomWoodward commented: 
yea i'm not sure the best way to organize it, ideally the heading levels would be based on some kind of semantic structure of the content sections, and i think that might exist in the xml but i'm not sure if it persists through to the html, at the time of the "web baking step" do we still have enough info to reliably re-level the headings?

@erikayvonne commented:
I ran a scan on the URL in this card [here](https://openstax.hub.essentia11y.com/scans/scan-results/6509aa814abef5b5ad5c734c?linkedPropertyData=64ca7a2b97d55500bbf06d9d%7C64adac862b178b2e09ae9e43). I'm seeing a couple of critical-severity issues, but heading structures skipping a level are not critical (it says that that manual review is required, but the rule also has low-severity). [Here's](https://openstax.hub.essentia11y.com/scans/scan-results/6509aa814abef5b5ad5c734c/aggregated-report/accessibility_access/402?linkedPropertyData=64ca7a2b97d55500bbf06d9d%7C64adac862b178b2e09ae9e43&pageNumber=1&pageSize=20&sortBy=&direction=&issueId=6509ab237b1d7236366d26db&dismissedFilter=&advancedSearch=) the issue:

> **What is the problem**
> Heading structure skips heading level
> 
> **Where is it located**
> 1.1 Functions and Function Notation - Precalculus 2e | OpenStax - https://openstax.org/books/precalculus-2e/pages/1-1-functions-and-function-notation
> 
> **Component name**
> < Not associated with a component >
>
> **XPath or CSS selector:**
> `body>div:nth-of-type(3)>*:nth-child(5)>*:nth-child(3)>*:nth-child(2)>*:nth-child(3)>*:nth-child(1)>*:nth-child(1)>*:nth-child(1)>*:nth-child(1)>*:nth-child(1)>*:nth-child(1)>*:nth-child(1)>*:nth-child(1)`
>
> **HTML Element:**
> `<h3 data-type="title">Learning Objectives</h3>`
> 
> **Description**
> One or more elements (created using h1 - h6 elements (without an ARIA-assigned role); or elements with a role attribute set to a value that starts with 'heading' and an aria-level attribute set to an appropriate value), not intentionally hidden in the DOM and available to assistive technologies, creates an inappropriate jump in heading levels within the document heading hierarchy.
> 
> **Why it matters**
> When levels are skipped, users of screen readers who are blind or low vision may not understand the relationship of sections of content.
> 
> **What to do**
> Use heading elements in the proper order on the page. For example, a heading level 1 (`h1`) generally indicates the primary purpose of the page and is followed by heading level 2 (`h2`). A heading level 2 (`h2h4h2`) section need not have any heading below it.
> 
> A rule of thumb is that in general, the heading level should match the visual representation of the heading level. It is possible to have a h2 followed by an h4 if the h4 (`h4`)represents content that is perhaps lower in structure or importance than a h3 where an h3 might be used elsewhere in the same article. However, if the `h4` is really the next logical level after `h2` and doesn't match an h4 visually from another part of the same article, or is not of lesser importance such as used in an error message or aside, then it should be an `h3` to communicate the structure.
> 
> Due to lack of support for the HTML5 sectional structure, heading need to align with the order of headings in the containing page and their corresponding HTML numbered heading levels (`h1`,`h2`, `h3`, etc.). ARIA `role="heading"` and `aria-level` can also be used to properly order the numbered heading levels in HTML5 outline structure.
> 
> **Findings History**
> First seen: September 19, 2023
> Last seen: September 19, 2023
> Previous status: "Open"
> Times seen: 1

[Here's](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships) the related article from understanding WCAG standards.

@openstaxalina commented: 
Thanks so much for the ping. I would say, given...

>  heading structures skipping a level are not critical (it says that that manual review is required, but the rule also has low-severity)

...that we **hold off on this until we have a separate baking step**. It seemed like anything else would cause a lot of breaking/testing/fixing of the PDFs (or it may not even be possible to get both formats working properly). @TomWoodward and/or @philschatz  please let me know if I missed an alternative, or misunderstood the information!

**Comment by lwtchu** [comment]
# Acceptance Criteria
- determine what's the best way to utilize the web bake step to address this issue
- the first non module title header should always be a H2 
- please check behavior on various pages - composite, introduction, appendix

**Comment by openstaxalina** [comment]
Thank you for the question and analysis, @kswanson33 . I think the best/most thorough fix you described is indeed best, because we don't want to fix the H2/H3 issue, only to create another header skipping issue further down the page. 

I can do any/all of the following if you and @lwtchu deem it helpful for confirming the plan:
- note for further discussion on the Thursday call agenda w/ DISCO
- send queries to Level Access via email (we have unlimited support, and they usually respond in 1-2 days)
- schedule a live call with Level Access to discuss this issue

Please let me know if any of these would be helpful.

**Comment by lwtchu** [comment]
TLDR whatever kendra said. to do. i agree.

**Comment by kswanson33** [comment]
Implementation plan:
- [ ] create direction to adjust headers
- [ ] create way for data that comes through cookbook to mock the structure of the data on rex (eg. only pages, no page title or metadata) & then replace original structure
- [ ] call header direction from web recipe

**Comment by kswanson33** [comment]
branch: cookbook `add-order-headers`

for QA: (copied from branch)

-  this impacts all headers in all web books
-  check that the headers in a range of textbook pages (preface, intro, regular page, eoc section, answer key, etc) progress linearly from H1 (outside of #main-content) to H2, H3, etc (inside #main-content) without skipping
-  styling will change a little for promoted headers

**Comment by MartaSztuka** [comment]
#### @kswanson33 I've tested precalculus and headers doesn;t progres by one.
![Image](https://github.com/openstax/cnx-recipes/assets/47241114/3286a0c3-b1c8-4321-a51b-a14a6e062398)
![Image](https://github.com/openstax/cnx-recipes/assets/47241114/cbb20f22-9d22-44be-82ce-89bc0528cf59)
#### and learning objectives should be H2
![Image](https://github.com/openstax/cnx-recipes/assets/47241114/cc6bdab3-8bbd-4318-80d5-03e6d2b9ec36)

**Comment by kswanson33** [comment]
college algebra is working for me locally:

<img width="985" alt="image" src="https://github.com/openstax/cnx-recipes/assets/26280712/36a9e026-4fa4-457c-9b71-fbaaf20bf7db">

<img width="988" alt="image" src="https://github.com/openstax/cnx-recipes/assets/26280712/6db51277-eaa9-4909-b3b9-b151f27faa76">

not sure what went wrong above ... it looks like the changes aren't being applied somehow. was this tested on the cookbook branch `add-order-headers`?

**Comment by Stackblocks** [comment]
When I test precalc with the cookbook `add-order-headers` branch I still get the skipped h2. I'll check College Algebra as well.


![Image](https://github.com/openstax/cnx-recipes/assets/12678235/617c8113-85be-42e0-8411-978578235b02)




![Image](https://github.com/openstax/cnx-recipes/assets/12678235/475f4e72-7ba3-4687-963d-0805946d2872)

**Comment by kswanson33** [comment]
thanks for the catch, updated the branch, hopefully this works 🤞

**Comment by Stackblocks** [comment]
The good news is that Algebra & Trigonometry now follows the logical header order. The bad news is that Precalculus 2e still doesn't.



![Image](https://github.com/openstax/cnx-recipes/assets/12678235/9d4e10c8-1673-46c0-add9-5bf2e758269f)



![Image](https://github.com/openstax/cnx-recipes/assets/12678235/8361fb15-0670-4e9b-b9a1-cb35d8c6f0b8)
